### PR TITLE
[CONTEXT-169] Add central state management

### PR DESCRIPTION
**Pull request by ghost**
Added in central stage management because it was getting messy.

Now there's no prop chaining!

### CODE REVIEW

File: src/containers/AssetViewer.tsx [hunk 1]
```diff
@@ -82,112 +83,96 @@ export class AssetViewer extends React.Component<Props, State> {
...
-            this.props.onAssetIdChange(rootAssetId, id)
-          }
         />
       </div>
     );
   };
 
   renderRelationshipsViewer = () => {
     const { rootAssetId } = this;
+    if (!rootAssetId) {
```
Comment (anderscognite) [inline src/containers/AssetViewer.tsx]: Should we only show it if user has access to relationships?

File: src/containers/Sidebar/SidebarNodeView.tsx [hunk 2]
```diff
@@ -3,65 +3,56 @@ import { connect } from 'react-redux';
...
-    const { modelId, revisionId, nodeId } = this.props;
-    const parent = await sdk.revisions3D.list3DNodes(modelId, revisionId, {
+    const { modelId, revisionId, nodeId, rootAssetId } = this.props.app;
+    const parent = await sdk.revisions3D.list3DNodes(modelId!, revisionId!, {
       nodeId,
     });
-    this.props.onNodeIdChange(parent.items[0].parentId);
+    this.props.setModelAndRevisionAndNode(
+      rootAssetId!,
+      parent.items[0].parentId
```
Comment (anderscognite) [inline src/containers/Sidebar/SidebarNodeView.tsx]: Does this always exist? Or can we have a zero length array?

File: src/store/middleware.ts [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+  setModelAndRevisionAndNode,
+} from '../modules/app';
+
+export const locationStateMiddleware: Middleware<
+  {},
+  RootState,
+  ThunkDispatch<any, any, SetAppStateAction>
+> = store => next => action => {
+  const response = next(action);
+  // Do stuff
```
Comment (anderscognite) [inline src/store/middleware.ts]: wat

File: src/store/middleware.ts [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    action.type === '@@router/LOCATION_CHANGE' &&
+    action.payload.type !== 'PUSH'
+  ) {
+    const { app } = store.getState();
+    const [, tenant, type, p1, p2, p3] = action.payload.location.pathname.split(
+      '/'
+    );
+    if (type === 'asset') {
+      const rootAssetId =
+        p1 && !Number.isNaN(parseFloat(p1)) ? Number(p1) : undefined;
```
Comment (anderscognite) [inline src/store/middleware.ts]: Are you sure this isNaN function always does what we think? See https://stackoverflow.com/questions/1303646/check-whether-variable-is-number-or-string-in-javascript for an additional requirement. I've had issues with this before :P

Also see https://stackoverflow.com/questions/25176459/is-number-isnan-more-broken-than-isnan

Comment (anderscognite) [reply src/store/middleware.ts]: Should we also extract this logic into a utils function as it is used many times?

**Review by anderscognite — APPROVED** [review]
Looks good! I have a couple of questions / comments I'd like you to check out / fix before merge :)

File: src/containers/AssetViewer.tsx (hunk 1 above)
Comment (dlcognite) [reply src/containers/AssetViewer.tsx]: Good point, we should add that in along with other permissions, we can track it here https://cognitedata.atlassian.net/browse/CONTEXT-257

File: src/store/middleware.ts (hunk 4 above)
Comment (dlcognite) [reply src/store/middleware.ts]: this is only done once so no need to extract

also the parsefloat should work the quirks 
we should not use `isNaN` (the global) as it is up to browser implementation, where as Number.isNaN is standardized.
```
Number.isNaN("asdfsad")
false
Number.isNaN(parseFloat("asdfsad"))
true
```

File: src/store/middleware.ts (hunk 3 above)
Comment (dlcognite) [reply src/store/middleware.ts]: opps, added useful comments

File: src/containers/Sidebar/SidebarNodeView.tsx (hunk 2 above)
Comment (dlcognite) [reply src/containers/Sidebar/SidebarNodeView.tsx]: added check

File: src/containers/AssetViewer.tsx (hunk 1 above)
Comment (dlcognite) [reply src/containers/AssetViewer.tsx]: tracking here
https://cognitedata.atlassian.net/browse/CONTEXT-257

we should also check on other permissions

**Pull request merged**